experiments/replicaset.go

```diff
@@ -0,0 +1,200 @@
+package experiments
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubernetes/pkg/controller"
+	labelsutil "k8s.io/kubernetes/pkg/util/labels"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/conditions"
+	experimentutil "github.com/argoproj/argo-rollouts/utils/experiment"
+	"github.com/argoproj/argo-rollouts/utils/log"
+	replicasetutil "github.com/argoproj/argo-rollouts/utils/replicaset"
+)
+
+var controllerKind = v1alpha1.SchemeGroupVersion.WithKind("Experiment")
+
+func (c *ExperimentController) getReplicaSetsForExperiment(experiment *v1alpha1.Experiment) (map[string]*appsv1.ReplicaSet, error) {
+	// List all ReplicaSets to find those we own but that no longer match our
+	// selector. They will be orphaned by ClaimReplicaSets().
+	rsList, err := c.replicaSetLister.ReplicaSets(experiment.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing ReplicaSets (see #42639).
+	canAdoptFunc := controller.RecheckDeletionTimestamp(func() (metav1.Object, error) {
+		fresh, err := c.arogProjClientset.ArgoprojV1alpha1().Experiments(experiment.Namespace).Get(experiment.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		if fresh.UID != experiment.UID {
+			return nil, fmt.Errorf("original Experiment %v/%v is gone: got uid %v, wanted %v", experiment.Namespace, experiment.Name, fresh.UID, experiment.UID)
+		}
+		return fresh, nil
+	})
+
+	templateToRS := make(map[string]*appsv1.ReplicaSet)
+	for _, template := range experiment.Spec.Templates {
+		replicaSetSelector, err := metav1.LabelSelectorAsSelector(template.Selector)
+		if err != nil {
+			return nil, fmt.Errorf("experiment %s/%s has invalid label selector: %v", experiment.Namespace, experiment.Name, err)
+		}
+		templateRSs := make([]*appsv1.ReplicaSet, 0)
+		for i := range rsList {
+			rs := rsList[i]
+			if replicaSetSelector.Matches(labels.Set(rs.ObjectMeta.Labels)) {
+				templateRSs = append(templateRSs, rs)
+			}
+		}
+
+		cm := controller.NewReplicaSetControllerRefManager(c.replicaSetControl, experiment, replicaSetSelector, controllerKind, canAdoptFunc)
+		templateRSs, err = cm.ClaimReplicaSets(templateRSs)
+		if err != nil {
+			return nil, err
+		}
+		matches := len(templateRSs)
+		if matches > 1 {
+			return nil, fmt.Errorf("multiple ReplicaSets match single experiment template")
+		} else if matches == 1 {
+			templateToRS[template.Name] = templateRSs[0]
+		}
+	}
+	return templateToRS, nil
+}
+
+func (ec *ExperimentController) getNewReplicaSet(experiment *v1alpha1.Experiment, template v1alpha1.TemplateSpec) (*appsv1.ReplicaSet, error) {
```

Comment:
I think `getNewReplicaSet` is misleading since the method might create/update replicaset. Probably `reconcileReplicaSet` is better name.